CRM/Sepa/Logic/Queue/Update.php

```diff
@@ -23,121 +23,182 @@
+        WHERE mandate.type = 'RCUR'
+          AND mandate.status = '$sdd_mode'
+          AND mandate.creditor_id = $creditor_id;");
+    }
+  }
+
+  /**
+   * Use Civi::queue to do the SDD group update
+   *
+   * @param $mode string
+   * @return void
    */
   public static function launchUpdateRunner($mode) {
-    if (!CRM_Sepa_Logic_Settings::acquireAsyncLock('sdd_async_update_lock', SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT)) {
-      CRM_Core_Session::setStatus(ts("Cannot run update, another update is in progress!", array('domain' => 'org.project60.sepa')), ts('Error'), 'error');
+    $sdd_async_update_lock = CRM_Sepa_Logic_Settings::acquireAsyncLock(
+      'sdd_async_update_lock',
+      SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT
+    );
+
+    if (!$sdd_async_update_lock) {
+      CRM_Core_Session::setStatus(
+        ts('Cannot run update, another update is in progress!', ['domain' => 'org.project60.sepa']),
+        ts('Error'),
+        'error'
+      );
+
       $redirect_url = CRM_Utils_System::url('civicrm/sepa/dashboard', 'status=active');
       CRM_Utils_System::redirect($redirect_url);
-      return; // shouldn't be necessary
+
+      // Shouldn't be necessary
+      return;
     }
-    // create a queue
-    $queue = CRM_Queue_Service::singleton()->create(array(
-      'type'  => 'Sql',
-      'name'  => 'sdd_update',
-      'reset' => TRUE,
-    ));
-
-    // first thing: close outdated groups
-    $queue->createItem(new CRM_Sepa_Logic_Queue_Update('PREPARE', $mode));
-    $queue->createItem(new CRM_Sepa_Logic_Queue_Update('CLOSE', $mode));
-
-    // then iterate through all creditors
-    $creditors = civicrm_api3('SepaCreditor', 'get', array('option.limit' => 0));
+
+    // Create a queue
+    $queue = Civi::queue('sdd_update', [
+      'error'  => 'delete',
+      'reset'  => TRUE,
+      'runner' => 'task',
+      'type'   => 'SqlParallel',
+    ]);
+
+    // Close outdated groups
+    $queue->createItem(self::createTask('PREPARE'));
+    $queue->createItem(self::createTask('CLOSE'));
+
+    // Then iterate through all creditors
+    $creditors = civicrm_api3('SepaCreditor', 'get', [ 'option.limit' => 0 ]);
+
     foreach ($creditors['values'] as $creditor) {
-      $sdd_modes = ($mode=='RCUR') ? array('FRST','RCUR') : array('OOFF');
+      $sdd_modes = ($mode == 'RCUR') ? ['FRST', 'RCUR'] : ['OOFF'];
+
       foreach ($sdd_modes as $sdd_mode) {
-        $count = self::getMandateCount($creditor['id'], $sdd_mode) + SDD_UPDATE_RUNNER_BATCH_SIZE; // safety margin
-        for ($offset=0; $offset < $count; $offset+=SDD_UPDATE_RUNNER_BATCH_SIZE) {
-          // add an item for each batch
-          $queue->createItem(new CRM_Sepa_Logic_Queue_Update('UPDATE', $sdd_mode, $creditor['id'], $offset, SDD_UPDATE_RUNNER_BATCH_SIZE));
+        // Safety margin
+        $count = self::getMandateCount($creditor['id'], $sdd_mode) + SDD_UPDATE_RUNNER_BATCH_SIZE;
+
+        for ($offset=0; $offset < $count; $offset += SDD_UPDATE_RUNNER_BATCH_SIZE) {
```

Comment:
This could have been a side-effect of the locking issue, but on dev2 I ran into a request timeout; I'm guessing enqueueing all the `UPDATE` tasks may be too much for a single HTTP request. We should figure out the locking issue first, but if we're still running into that issue afterwards, we may need to add a kind of meta-queue that adds these in batches instead.